### DISCUSSION TITLE
Gestion liaison multiple pour une valeur

### DISCUSSION THREAD

**Discussion by Beetix** [witsa/synapps#736] [Q&A] [answered]
Bonjour,

Aurait-il une bonne manière pour gérer une valeur qui en dépend de plusieurs ? Par exemple, pour un composite dont il faut changer la couleur en fonction de son état de marche et de défaut. La solution que j'utilise actuellement est d'avoir des `onValueChanged` sur marche et défaut avec le même script modifiant la couleur. Cela n'est pas idéal car le code est dupliqué...

**Comment by JeremieDENAEGEL** [comment]
Bonjour,

Pour ma part, j'utilise actuellement 2 méthodes suivant les cas.
La première: Côté ressource, je crée un évaluateur analogique avec des Select pour obtenir un chiffre différent suivant les cas. Puis coté synapps, je fais du OnReadTransform avec du switch case sur le résultat de l'évaluateur.

La seconde: (peut-être plus réfléchi si on pense en synapps pour highway), je fais un additionnel pour chaque élément et j'y mets un script OnPropertychanged avec des "if else" suivant la logique. 

Il y a probablement d'autres méthodes possible

**Reply by Beetix** [reply]
Merci pour le retour.

Nous essayons au maximum de ne pas rajouter des ressources non métier ce qui fait que nous n'avons pas envisagé la première méthode.

Concernant la seconde, c'est justement ce que nous avons fait. Le souci étant que le code est dupliqué dans chaque `onValueChanged`. Après nous ne sommes pas allés au bout, mais il y a peut-être moyen de limiter la duplication en encapsulant le code dans une fonction rattachée à l'acteur appelée à chaque événement.

**Comment by manufitoussiwit** [comment, marked as answer]
Salut,

Il y a plusieurs façon de réaliser ça sans duplication de code. Y en a deux que j'aime bien : 

# avec `onPropertyChanged`

Le principe est de créer autant d'additionnelle sur l'acteur qui doit changer que de valeurs qui vont l'influancer. Par exemple ici pour un acteur Texte qui dois changer de couleur en fonction d'une valeur d'état on/off et s'il est en faute.
J'ai créé deux additionnelles de type Booléen : une pour l'état on/off et une autre pour la faute.
Ces deux additionnelles seront liées à la donnée.
Ensuite, je crée dans l'évènement `onPropertyChanged` un script qui va detecter quand l'une de ces deux additionnelle change de valeur pour appliquer la couleur résultante à l'acteur.

![demo-multi-additionnals1](https://github.com/witsa/synapps/assets/35595723/3153fd2c-4c48-4cd0-aef7-766b320197ad)


```js
if ([
  'additionals.onOff',
  'additionals.fault',
].includes(context.propertyPath)) {
  if (this.additionals.fault) {
    this.properties.color = 'red';
  } else {
    this.properties.color = this.additionals.onOff ? 'green' : 'black';
  }
}
```

La scène exemple 1 :
```
SYNAPPS-STUDIO-SCENE|{"config":{"key":"scene132","name":"Exemple 1"},"leadActor":{"type":"layout/stack","key":"stack1","children":[{"type":"display/text","key":"text1","properties":{"verticalAlignment":"middle","horizontalAlignment":"middle","content":"Exemple","fontSize":"6vmin"},"additionalDefs":{"onOff":{"type":"boolean"},"fault":{"type":"boolean"}},"events":{"onPropertyChanged":["if ([","  'additionals.onOff',","  'additionals.fault',","].includes(context.propertyPath)) {","  if (this.additionals.fault) {","    this.properties.color = 'red';","  } else {","    this.properties.color = this.additionals.onOff ? 'green' : 'black';","  }","}"]}}]}}
```

# avec `onRender` ou `onStyle

L'autre méthode consiste toujours à ajouter les additionnelles mais cette fois, on indique lors de leur création le type modification qu'elles impliquent si leur valeur change : on peut choisir *Style CSS* ou *Rendu*. C'est en fonction de ce que ça implique vraiment. Ici, *Style CSS* suffirait mais pour être général je choisis *Rendu*.

![image](https://github.com/witsa/synapps/assets/35595723/31d84b4d-ce9b-4369-9370-bd44c357494c)

Ensuite, au lieu de faire un script dans `onPropertyChanged`, on le fait dans `onRender` (pour *Style CSS* c'est `onCssStyling` : 

```js
  if (this.additionals.fault) {
    this.properties.color = 'red';
  } else {
    this.properties.color = this.additionals.onOff ? 'green' : 'black';
  }
```

On voit qu'on a plus besoin de filtrer les cas où c'est bien les additionnelles qui ont changé.

La scène exemple 2 :
```
SYNAPPS-STUDIO-SCENE|{"config":{"key":"scene132","name":"Exemple 2"},"leadActor":{"type":"layout/stack","key":"stack1","children":[{"type":"display/text","key":"text1","properties":{"verticalAlignment":"middle","horizontalAlignment":"middle","content":"Exemple","fontSize":"6vmin"},"additionalDefs":{"onOff":{"type":"boolean","modifier":"render"},"fault":{"type":"boolean","modifier":"render"}},"events":{"onRender":["  if (this.additionals.fault) {","    this.properties.color = 'red';","  } else {","    this.properties.color = this.additionals.onOff ? 'green' : 'black';","  }"]}}]}}
```

> Perso j'ai une préférence pour la première méthode. Il y a plus de contrôle.

Pour finir, un lien vers la doc qui parle des évènements : https://witsa.github.io/synapps/concepts/scripts/actor-life-cycle.html#changement-de-valeur-de-propri%C3%A9t%C3%A9-dun-acteur

Bien cordialement.

**Reply by Beetix** [reply]
Merci pour ces 2 solutions et les exemples !

Je viens de tester le `onRender` et cela fonctionne très bien.

L'explication du `onPropertyChanged` m'a permis de comprendre la solution initialement proposée par @JeremieDENAEGEL qui est tout aussi pertinente.
